returnn/tf/util/basic.py

```diff
@@ -250,26 +250,44 @@ def set_padding_info(x, dim, pad_value):
 def mask_dyn_seq_len_nd(x, pad_value, axes):
   """
   :param Data x:
-  :param float|int pad_value:
+  :param float|int|tf.Tensor pad_value:
   :param list[int]|tuple[int] axes:
   :return: masked x
   :rtype: tf.Tensor
   """
+  # Filter out some axes which should not be used for masking.
+  axes_ = []
+  for axis in axes:
+    tag = x.dim_tags[axis]
+    assert tag.dyn_size_ext
+    # It only makes sense to apply for this axis if the dyn size dims are all existing in x itself.
+    # E.g. if the dyn_size_ext shape is [B] but the shape of x is just [T] (without B),
```

Comment:
Wait, isn't this very dangerous / wrong?
In this case, you'd want the output of the layer to have this implicit dims (`B` in your example) as concrete dims in the output. As in https://github.com/rwth-i6/returnn/issues/721. Right?
We e.g. had a bug with this before in https://github.com/rwth-i6/returnn/issues/717.

Reply:
The usage of `mask_dyn_seq_len_nd` is only in `DotLayer` and in `Data.copy_masked` which itself is only used for the ext dyn size in `Data.get_sequence_mask_broadcast`. 

In `DotLayer`, we already check for that.

In `Data.get_sequence_mask_broadcast` for the ext dyn size, it should not matter. We just want to make sure that there are no invalid sizes (zero sizes).

You are right, it could be wrong in other cases. Currently there are no other cases and I think the current cases are both correct.

So, what should we do? Maybe just add a big warning comment with this info there?